### PR TITLE
[9.0][ADD] Customer Outstanding Statement

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Print Customer Outstanding Statement
====================================

The outstanding statement provides details of all outstanding customer receivables
up to a particular date. This includes all unpaid invoices, unclaimed refunds and
outstanding payments. The list is displayed in chronological order and is split by currencies.

Aging details can be shown in the report, expressed in aging buckets (30 days due, ...), 
so the customer can review how much is open, due or overdue.

Configuration
=============

Users willing to access to this report should have proper Accounting & Finance rights:

1. Go to *Settings / Users* and edit your user to add the corresponding access rights as follows.
2. In *Application / Accounting & Finance*, select *Accountant* or *Adviser* options.

Usage
=====

To use this module, you need to:

1. Go to Customers and select one or more
2. Press 'Action > Customer Outstanding Statement Report'
3. Indicate if you want to display aging buckets

### CODE REVIEW

File: customer_outstanding_statement/README.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+====================================
+Print Customer Outstanding Statement
+====================================
+
+The outstanding statement provides details of all outstanding on the customer receivables
```
Comment (elicoidal) [inline customer_outstanding_statement/README.rst]: The outstanding statement provides details of all outstanding customer receivables entries
up to a particular date

File: customer_outstanding_statement/README.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+up until a particular date. This includes all unpaid invoices, unclaimed refunds and
+outstanding payments. The list is displayed in chronological order and is split by currencies.
+
+Aging details can be shown in the report, expressed in aging buckets (30 days
+due, ...), so the customer can review how much is open, due or overdue.
+
```
Comment (elicoidal) [inline customer_outstanding_statement/README.rst:15]: A screenshot of the report would be welcome

File: customer_outstanding_statement/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#   (http://www.eficent.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'Customer Outstanding Statement',
+    'version': '9.0.1.0.0',
+    'category': 'Reports/pdf',
+    'summary': 'OCA Financial Reports',
+    'author': "Eficent, Odoo Community Association (OCA)",
+    'website': 'https://github.com/OCA/account-financial-reporting',
```
Comment (elicoidal) [inline customer_outstanding_statement/__openerp__.py:12]: usually the website of the author or OCA

File: customer_outstanding_statement/report/customer_outstanding_statement.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Eficent Business and IT Consulting Services S.L.
+#   (http://www.eficent.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from datetime import datetime, timedelta
+from openerp.tools import DEFAULT_SERVER_DATE_FORMAT
```
Comment (elicoidal) [inline customer_outstanding_statement/report/customer_outstanding_statement.py:7]: I think you can replace all openerp by odoo

File: customer_outstanding_statement/views/statement.xml [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                                    <td class="text-right">
+                                        <span t-esc="Amount_Due[o.id][currency]" t-esc-options='{"widget": "monetary", "display_currency": "currency"}'/>
+                                    </td>
+                                </tr>
+                            </table>
+                            <table class="table table-condensed" border="1" t-if="Show_Buckets">
+                                <thead>
+                                    <tr>
+                                        <th class="text-center">Current Due</th>
+                                        <th class="text-center">1-30 Days Due</th>
```
Comment (elicoidal) [inline customer_outstanding_statement/views/statement.xml]: not sure here whether Days or Day

File: customer_outstanding_statement/wizard/customer_outstanding_statement_wizard.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    date_end = fields.Date(required=True,
+                           default=fields.Date.to_string(date.today()))
+    show_aging_buckets = fields.Boolean(string='Include Aging Buckets',
+                                        default=True)
+    number_partner_ids = fields.Integer(
+        default=lambda self: len(self._context['active_ids'])
+    )
+    filter_partners_non_due = fields.Boolean(
+        string='Don\'t show partners that have nothing due', default=True)
```
Comment (elicoidal) [inline customer_outstanding_statement/wizard/customer_outstanding_statement_wizard.py]: 'Don\'t show partners with no due entries'

File: customer_outstanding_statement/views/statement.xml [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                                    </td>
+                                    <td class="text-right">
+                                        <span t-esc="bucket['balance']" t-esc-options='{"widget": "monetary", "display_currency": "currency"}'/>
+                                    </td>
+                                </tr>
+                            </table>
+                        </t>
+                    </t>
+                    <p t-if="not Lines[o.id]">
+                        <strong>There is nothing due with this partner.</strong>
```
Comment (elicoidal) [inline customer_outstanding_statement/views/statement.xml]: s/with/for

**Review by elicoidal — CHANGES_REQUESTED** [review]
some details, most of them non-blocking

File: customer_outstanding_statement/README.rst [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+Print Customer Outstanding Statement
+====================================
+
+The outstanding statement provides details of all outstanding customer receivables
+up to a particular date. This includes all unpaid invoices, unclaimed refunds and
+outstanding payments. The list is displayed in chronological order and is split by currencies.
+
+Aging details can be shown in the report, expressed in aging buckets (30 days
+due, ...), so the customer can review how much is open, due or overdue.
+
```
Comment (jcdrubay) [inline customer_outstanding_statement/README.rst:15]: Could we add something like:
```
Note: this report is a more detailed version of the native report "Due payments" which is available from the partner form: Print > Due Payments
```

**Review by jcdrubay — APPROVED** [review]
Reviewed the code and tested on Runbot. Very nice work. However, I think most of companies would be satisfied with the native Due Payments report; we should help them to choose wisely.

Approved assuming comments from @elicoidal are taken into account.

File: customer_outstanding_statement/__openerp__.py (hunk 3 above)
Comment (MiquelRForgeFlow) [reply customer_outstanding_statement/__openerp__.py:12]: We have seen that the usual practice is to put the URL where the module can be found.
cc: @pedrobaeza

File: customer_outstanding_statement/report/customer_outstanding_statement.py (hunk 4 above)
Comment (MiquelRForgeFlow) [reply customer_outstanding_statement/report/customer_outstanding_statement.py:7]: This should be fine in version 10.0, but not in the 9.0 :nerd_face:

File: customer_outstanding_statement/README.rst (hunk 2 above)
Comment (MiquelRForgeFlow) [reply customer_outstanding_statement/README.rst:15]: Screenshot added 😄

File: customer_outstanding_statement/README.rst (hunk 8 above)
Comment (MiquelRForgeFlow) [reply customer_outstanding_statement/README.rst:15]: The Due Payments module (<code>account.report_overdue</code>) you have referred it prints all activity on the customer receivables from all the history of the customer, nothing related with the Customer Outstanding Statement that just prints the outstanding. The Customer Activity Statement (https://github.com/OCA/account-financial-reporting/pull/290) is more related to the Due Payments module.

**Comment by JordiBForgeFlow** [comment]
@mreficent when you open the wizard there's a text 

> Customer Outstanding Statement is a detailed report of the receivables by intervals of time.

This should be changed to:

> The outstanding statement provides details of all outstanding customer receivables
up to a particular date. This includes all unpaid invoices, unclaimed refunds and
outstanding payments. The list is displayed in chronological order and is split by currencies.
> Aging details can be shown in the report, expressed in aging buckets (30 days due, ...),
so the customer can review how much is open, due or overdue.

**Comment by MiquelRForgeFlow** [comment]
@jbeficent Done :slightly_smiling_face:

**Review by JordiBForgeFlow — APPROVED** [review]

File: customer_outstanding_statement/report/customer_outstanding_statement.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                                sum(b_60_90) as b_60_90,
+                                sum(b_90_120) as b_90_120,
+                                sum(b_over_120) as b_over_120
+            FROM Q3
+            GROUP BY partner_id, currency_id
+        """
+
+    _bucket_dates = {
+        'today': fields.date.today(),
+        'minus_30': fields.date.today() - timedelta(days=30),
```
Comment (elicoidal) [inline customer_outstanding_statement/report/customer_outstanding_statement.py:219]: A really killing nice to have (and not so difficult IMO) would be to be able to set up the time buckets

**Review by elicoidal — APPROVED** [review]
Great

**Pull request merged**